### ISSUE TITLE
Whirlwind Barbarian won't attack Undead Physical immunes

### ISSUE THREAD

**Issue by Karakachanov** [kolton/d2bot-with-kolbot#2433]
Got whirlwind barbarian with Azurewrath (sanctuary aura that removes undead physical immunities ), But the bot won't recognize undead as non immune and tries to attack them with the alternative skill (Berserk).

I'm trying to make ghosthunter barb, but using only berserk is not practical and too slow. Tried to set up profile without alternative attack skill , or with alternative attack skill that's still whirlwind or with custom attack skill set up only for ghost types:

"ghost":[151],"wraith":[151],"specter":[151],"apparition":[151],"darkshape":[151]

It doesn't matter what i do, bot always ignores undead physical immunes or attacks them only with berserk.

Is there some way to make whirlwind work against immunes(even when it can't damage them), or change specific monsters so bot considers them non-immune ?

**Comment by DarkHorseDre** [comment]
I think the bot checks mob immunity and your skills before attacking, and also checks for infinity aura, and will attack with berserk or avoid physical immunes.  I'm guessing (cant check code right now) that the bot simply isn't checking sanctuary aura or indeed the resultant lower physical immunity.

I think the functions in attack.js are canAttack and checkRes - if you know how to alter it you could force it to check for azurewrath/santuary or simply fix why its not reading the reduced immunity of the mob (eg. attack.js checks for conviction aura but not sanctuary)

**Comment by Karakachanov** [comment]
Yes, thanks for your reply. I did check attack.js and i think i know where the problem is:

	// Check if a monster is immune to specified attack type
	checkResist: function (unit, val, maxres) {
		// Ignore player resistances
		if (unit.type === 0) {
			return true;
		}

		var damageType = typeof val === "number" ? this.getSkillElement(val) : val;

		if (maxres === undefined) {
			maxres = 100;
		}

		// Static handler
		if (val === 42 && this.getResist(unit, damageType) < 100) {
			return (unit.hp * 100 / 128) > Config.CastStatic;
		}

		if (this.infinity && ["fire", "lightning", "cold"].indexOf(damageType) > -1 && unit.getState) { // baal in throne room doesn't have getState
			if (!unit.getState(28)) {
				return this.getResist(unit, damageType) < 117;
			}

			return this.getResist(unit, damageType) < maxres;
		}

		return this.getResist(unit, damageType) < maxres;
	},

I'm not at all experienced with code but from what i see, there is exception built in for when you have infinity, but no exception for sanctuary aura vs undead (either with azurewrath/lawbringer or just paladin using sanctuary aura as a skill). Game itself doesn't change the status of undead monsters when using sanctuary, so they remain flagged as immune (what sanctuary does is to remove the built in physical resistance (i think it's 50%) that sometimes stacks with monster's actual physical resist to make it immune, that's why most mace type weapons (staves,wands,scepters) deal +50% damage to undead monsters, but they don't eliminate the immunity.)

**What would be the way to make such exception(like the one for infinity) for azurewrath/lawbringer but only vs undead monsters ?** (to my knowledge there are no undead physical immunities that sanctuary aura doesn't break, but I may be wrong)

This kind of build WW barb + azurewrath is very useful for the "ghostbuster" routine , since he can "find item" corpses in solo games and get drops based on his's skill and not on the number of players in the game.

**Comment by DarkHorseDre** [comment]
it is the damage res it reduces by 50% but physical is reduced to 0 - https://diablo.gamepedia.com/Sanctuary_(Diablo_II)

to answer, this is simple to do in isolation, but to work with the existing code will require understanding of how the code works, which I don't have fully.

from what I see the bot checks the immunises and the skills you have, and then applies assumptions - which do not include any appreciation for physical resistance reduction from the aforementioned items, or azurewrath more specifically.

So you could copy how the bot checks for infinity, and check for azurewrath/sanctuary aura (attack.js checks for state 28 which is the state for the conviction aura from infinity (sanctuary aura state is 47), then add code to have it check for WW skill (and others?) and if so, attack anyway.  atm it checks light/fire/cold and ignores physical.

Hopefully this is all you need to do - let me know if you get stuck and I'll have a look if/when I get time

**Comment by DarkHorseDre** [comment]
example:  This is the infinity check
```js
// Check player infinity
		item = me.getItem(-1, 1);

		if (item) {
			do {
				if (item.getPrefix(20566)) {
					this.infinity = true;

					return true;
				}
			} while (item.getNext());
		}

		return false;
	},
```

azurewraths id is 2603 so you could create a new function and change:

```js
if (item.getPrefix(2603)) {
					this.azurewrath = true;
```
and 
```js
if (!unit.getState(47)) {
				return this.getResist(unit, damageType); //remove 117 test as you know you'll break immunity for undead, but you may want to test that
			}
```

**Comment by Karakachanov** [comment]
it doesn't seem to work, or I'm not doing it right. 

Also a problem is sanctuary aura doesn't break all immunities, only undead ones, and this kind of exception (if i manage to make it) doesn't account for enemy being undead or animal or demon.

Infinity breaks monster immunities regardless of their type, and ingame flag for immune disappears(under the name of the monster) while with sanctuary aura undead monsters still keep their physical immune flag(ghosts are still labeled physical immune, even tho i can damage them, i guess that's a bug). I'm not sure how the program checks the enemy resist...

**Comment by DarkHorseDre** [comment]
no its ok - I think you just attack them anyway, where monster type is undead

I think gamedata.js specifies types you can test, as attack.js does:
```js
case "holybolt": // check if a monster is undead
			if (getBaseStat("monstats", unit.classid, "lUndead") || getBaseStat("monstats", unit.classid, "hUndead")) {
```

so I guess you have:

if mob is of undead types
and if using azurewrath (getunit) OR santuary aura affecting mob (dont think level of aura counts)
and if WW skill (getskill)
then attack with ww blindly as you know it will hurt em, no?

**Comment by DarkHorseDre** [comment]
> ```js
> if (!unit.getState(47)) {

to be clear, this is checking to see if mob is under aura - if the aura works as expected, then you know phys res is 0 and therefore you can override the code to force barb with WW to attack
